core/src/Streamly/Internal/Data/MutArray/Type.hs

```diff
@@ -2086,24 +2092,30 @@ fromListRev :: (MonadIO m, Unbox a) => [a] -> m (MutArray a)
 fromListRev xs = fromListRevN (Prelude.length xs) xs
 
 -------------------------------------------------------------------------------
--- Combining
+-- Cloning
 -------------------------------------------------------------------------------
 
--- | Put a sub range of a source array into a subrange of a destination array.
--- This is not safe as it does not check the bounds.
-{-# INLINE putSliceUnsafe #-}
-putSliceUnsafe :: MonadIO m => MutArray a -> Int -> MutArray a -> Int -> Int -> m ()
-putSliceUnsafe src srcStartBytes dst dstStartBytes lenBytes = liftIO $ do
-    assertM(lenBytes <= arrBound dst - dstStartBytes)
-    assertM(lenBytes <= arrEnd src - srcStartBytes)
-    let !(I# srcStartBytes#) = srcStartBytes
-        !(I# dstStartBytes#) = dstStartBytes
-        !(I# lenBytes#) = lenBytes
-    let arrS# = getMutableByteArray# (arrContents src)
-        arrD# = getMutableByteArray# (arrContents dst)
-    IO $ \s# -> (# copyMutableByteArray#
-                    arrS# srcStartBytes# arrD# dstStartBytes# lenBytes# s#
-                , () #)
+{-# INLINE cloneAs #-}
+cloneAs :: MonadIO m => PinnedState -> MutArray a -> m (MutArray a)
+cloneAs ps src =
+    liftIO $ do
+        let startSrc = arrStart src
+            srcLen = arrEnd src - startSrc
+        newArrContents <- Unboxed.newBytesAs ps srcLen
+        putSliceUnsafe (arrContents src) startSrc newArrContents 0 srcLen
+        return $ MutArray newArrContents 0 srcLen srcLen
+
+{-# INLINE clone #-}
+clone :: MonadIO m => MutArray a -> m (MutArray a)
+clone = cloneAs Unpinned
```

Comment:
Is this operation faster than streaming the array and writing it to a new array? Why do we need this?

Reply:
I'll need to check if it's faster.

Reply:
See https://github.com/composewell/streamly/issues/2573